### DISCUSSION TITLE
How to solve the bivariate LMI problem?

### DISCUSSION THREAD

**Discussion by yunjiandaiju** [yalmip/YALMIP#1337] [Problems using YALMIP] [answered]
Recently have been reading an article where I need to solve for a bivariate LMI. i have not been able to get the right answer, hopefully this can help me. Article name: From Noisy Data to Feedback Controllers:Nonconservative Design via a Matrix S-Lemma.Here is the code I used.
clear, clc;
% oo=1;
% EE=cell(100,100);
% cc=zeros(100,100);
% while(1)
%状态空间
A = [0.850 -0.038 -0.380;
0.735 0.815 1.594;
-0.664 0.697 -0.064];
B = [1.431 0.705;
1.620 -1.129;
0.913 0.369];
u = randn(2,20);
x0 = randn(3, 1);
%% 生成噪音
n = 3; % 噪声向量的长度
num_samples = 100; % 要生成的噪声向量数量
desired_norm_square = 0.5; %限制范数界
scaled_noises = zeros(n, num_samples);
for i = 1:num_samples
noise = randn(n, 1); % 生成高斯分布的随机噪声
while norm(noise, 2)^2 > desired_norm_square
noise = randn(n, 1); % 如果范数平方超过了阈值，重新生成噪声
end
scaled_noises(:, i) = noise;
end
% 从生成的噪声中随机抽取20列
num_selected = 20;
selected_indices = randperm(num_samples, num_selected);
w= scaled_noises(:, selected_indices);
%% 生成响应
t=20;x=[];x(:,1)=x0;
for i=1:t
x(:,i+1)=Ax(:,i)+Bu(:,i)+w(:,i);
end
x_i0=x(:,1:20);
x_i1=x(:,2:21);
U_i=u;
%% 求解LMI
%定义给定矩阵
I_3=eye(3);
I_0=zeros(3,3);
I_01=zeros(2,3);
I_02=zeros(1,20);
I_03=zeros(3,20);
I_04=zeros(20,3);
I_05=zeros(3,2);
I_06=zeros(2,3);
I_07=zeros(2,2);
a=randn;
b=randn;
fai_11=t0.5eye(3);
fai_22=-eye(20);
%% 定义N
A1=[I_3,x_i1;I_0,-x_i0;I_01,-U_i;I_0,I_03];
A2=[fai_11 I_03;I_04 fai_22];
N=aA1A2A1’;
[F2,G2]=eig(N);
G2;
N11=N(1:3,1:3);
N12=N(1:3,4:6);
N13=N(1:3,7:8);
N14=N(1:3,9:11);
N21=N(4:6,1:3);
N22=N(4:6,4:6);
N23=N(4:6,7:8);
N24=N(4:6,9:11);
N31=N(7:8,1:3);
N32=N(7:8,4:6);
N33=N(7:8,7:8);
N34=N(7:8,9:11);
N41=N(9:11,1:3);
N42=N(9:11,4:6);
N43=N(9:11,7:8);
N44=N(9:11,9:11);
% n = 23; % 方阵的大小
% N1 = zeros(n);
% % 设置左上角为3x3单位阵
% N1(1:3, 1:3) = 200.5eye(3);
% % 设置右下角为3x3单位阵
% N1(n-2:n, n-2:n) = -eye(3);
% N=aA1N1A1’;
%% 定义决策变量
% %CVX工具箱
cvx_begin
cvx_solver sdpt3
variable P(3, 3) semidefinite
variable L(2, 3)
minimize trace(P)
subject to
% [P-(bI_3)-N11,I_0-N12,I_05-N13,I_0-N14;I_0-N21,-P-N22,-(L’)-N23,I_0-N24;I_06-N31,-L-N32,I_07-N33,L-N34; I_0-N41,I_0-N42,(L’)-N43,P-N44]==semidefinite(11);% 半正定约束
[[P-(bI_3),I_0,I_05,I_0;I_0,-P,-(L’),I_0;I_06,-L,I_07,L;I_01,I_0,(L’),P]-N]==semidefinite(11);% 半正定约束
cvx_end
%定义优化问题
k=Linv(P);
U=A+(Bk);
[F1,E]=eig(U);
E
% EE{oo,oo}=E;
% oo=oo+1;
% if oo==100
% break;
% end
% end
% for i=1:100
% if any(max(EE{i,i})>=1)
% cc(i,i)=1;
% end
% end
% a=sum(cc(:));
% disp(a);
![20231008161840](https://github.com/yalmip/YALMIP/assets/142418156/308d03cf-c4d2-4135-9d66-42e27bb15427)
The picture shows the LMI to be solved, where P and L are variables. p is a real symmetric semipositive definite matrix and L is unbounded.I can work out the solution every time, but I think he's wrong. I brought the solution back into the system equations and could not verify that the system was stable. I hope I can get some help with this.

**Comment by johanlofberg** [comment, marked as answer]
This is a YALMIP forum, not a CVX forum.
https://yalmip.github.io/tutorial/basics/

Not clear why you are talking about bivariate, as it appears linear since your only decision variables are P and L

**Reply by yunjiandaiju** [reply]
Yes, I know it's yalmip forum, I talk about binary variables is that I may simply think that he contains two variables (P,L), which is the error of my theory. But I don't require him to be convex, so it shouldn't be a CVX problem. I'm confused about this.

**Reply by yunjiandaiju** [reply]
I'm very sorry that my program was uploaded wrong, I tried CVX solving before but none of them yielded results.
 clear, clc;
%  oo=1;
%  EE=cell(100,100);
%  cc=zeros(100,100);
% while(1)
%状态空间
A = [0.850 -0.038 -0.380; 
    0.735 0.815   1.594; 
    -0.664 0.697  -0.064];
B = [1.431 0.705;
    1.620 -1.129; 
    0.913 0.369];
u = randn(2,20);
x0 = randn(3, 1);
%% 生成噪音
n = 3; % 噪声向量的长度
num_samples = 100; % 要生成的噪声向量数量
desired_norm_square = 0.5; %限制范数界
scaled_noises = zeros(n, num_samples);
for i = 1:num_samples
    noise = randn(n, 1); % 生成高斯分布的随机噪声
    while norm(noise, 2)^2 > desired_norm_square
        noise = randn(n, 1); % 如果范数平方超过了阈值，重新生成噪声
    end
    scaled_noises(:, i) = noise;    
end
% 从生成的噪声中随机抽取20列
num_selected = 20;
selected_indices = randperm(num_samples, num_selected);
w= scaled_noises(:, selected_indices);
%% 生成响应
t=20;x=[];x(:,1)=x0;
for i=1:t
    x(:,i+1)=A*x(:,i)+B*u(:,i)+w(:,i);
end
x_i0=x(:,1:20);
x_i1=x(:,2:21);
U_i=u;
%% 求解LMI
%定义给定矩阵
I_3=eye(3);
I_0=zeros(3,3);
I_01=zeros(2,3);
I_02=zeros(1,20);
I_03=zeros(3,20);
I_04=zeros(20,3);
I_05=zeros(3,2);
I_06=zeros(2,3);
I_07=zeros(2,2);
a=randn;
b=randn;
fai_11=t*0.5*eye(3);
fai_22=-eye(20); 
%% 定义N
A1=[I_3,x_i1;I_0,-x_i0;I_01,-U_i;I_0,I_03];
A11=[I_3,x_i1;I_0,-x_i0;I_01,-U_i];
A2=[fai_11 I_03;I_04 fai_22];
N=a*A1*A2*A1';
N_11= A11*A2*A11';
[F2,G2]=eig(N);
G2;
N11=N(1:3,1:3);
N12=N(1:3,4:6);
N13=N(1:3,7:8);
N14=N(1:3,9:11);
N21=N(4:6,1:3);
N22=N(4:6,4:6);
N23=N(4:6,7:8);
N24=N(4:6,9:11);
N31=N(7:8,1:3);
N32=N(7:8,4:6);
N33=N(7:8,7:8);
N34=N(7:8,9:11);
N41=N(9:11,1:3);
N42=N(9:11,4:6);
N43=N(9:11,7:8);
N44=N(9:11,9:11);
% n = 23; % 方阵的大小
% N1 = zeros(n);
% % 设置左上角为3x3单位阵
% N1(1:3, 1:3) = 20*0.5*eye(3);
% % 设置右下角为3x3单位阵
% N1(n-2:n, n-2:n) = -eye(3);
% N=a*A1*N1*A1';
%% 定义决策变量
P = sdpvar(3,3);
L = sdpvar(2,3);
P2 =P-(b*I_3);
%定义优化问题
%约束LMI
L1 = [[  P2-N11        I_0-N12   I_05-N13   I_0-N14;
          I_0-N21       -P-N22    -(L')-N23   I_0-N24;
          I_06-N31      -L-N32    I_07-N33    L-N34; 
          I_0-N41       I_0-N42   (L')-N43    P-N44]>=0];
L2 = [P>=0]
LL =  L1+L2;
solvesdp(LL);
 %% 输出求解值
P1 = double(P);
[F,G]=eig(P1);
G
k = double(L)*inv(double(P))
U=A+(B*k);
[F1,E]=eig(U);
E
%   EE{oo,oo}=E;
% oo=oo+1;
%  if oo==100
%    break;
% end
%  end
% for i=1:100
%    if any(max(EE{i,i})>=1)
%        cc(i,i)=1;
%    end
%  end
% a=sum(cc(:));
% disp(a);

**Reply by johanlofberg** [reply]
If you only have P and L, it is convex so either CVX or YALMIP work just as well. If you want to have support here, you will have to supply YALMIP code (and put your code in a code block, otherwise it gets messed up as you can see on your code)

**Reply by yunjiandaiju** [reply]
Yes, I have re-uploaded the code. My goal was to solve for the LMI in the image, but I have not been able to get satisfactory results, I think it is a problem with my use of the software.

**Reply by johanlofberg** [reply]
no, the code is broken as you haven't placed it in a code block

**Reply by yunjiandaiju** [reply]
[xitongwending.zip](https://github.com/yalmip/YALMIP/files/12843216/xitongwending.zip)
Sorry, I don't particularly understand what you mean. Putting the code in a zip?

**Reply by johanlofberg** [reply]
You use the <> block that is above the text box to format

**Reply by johanlofberg** [reply]
After you created, look at preview, copy to matlab and make sure it runs

**Reply by yunjiandaiju** [reply]
```
 clear, clc;

A = [0.850 -0.038 -0.380; 
    0.735 0.815   1.594; 
    -0.664 0.697  -0.064];
B = [1.431 0.705;
    1.620 -1.129; 
    0.913 0.369];
u = randn(2,20);
x0 = randn(3, 1);

n = 3; % 
num_samples = 100; % 
desired_norm_square = 0.5; %
scaled_noises = zeros(n, num_samples);
for i = 1:num_samples
    noise = randn(n, 1); 
    while norm(noise, 2)^2 > desired_norm_square
        noise = randn(n, 1); 
    end
    scaled_noises(:, i) = noise;    
end
num_selected = 20;
selected_indices = randperm(num_samples, num_selected);
w= scaled_noises(:, selected_indices);
t=20;x=[];x(:,1)=x0;
for i=1:t
    x(:,i+1)=A*x(:,i)+B*u(:,i)+w(:,i);
end
x_i0=x(:,1:20);
x_i1=x(:,2:21);
U_i=u;
I_3=eye(3);
I_0=zeros(3,3);
I_01=zeros(2,3);
I_02=zeros(1,20);
I_03=zeros(3,20);
I_04=zeros(20,3);
I_05=zeros(3,2);
I_06=zeros(2,3);
I_07=zeros(2,2);
a=randn;
b=randn;
fai_11=t*0.5*eye(3);
fai_22=-eye(20); 
A1=[I_3,x_i1;I_0,-x_i0;I_01,-U_i;I_0,I_03];
A11=[I_3,x_i1;I_0,-x_i0;I_01,-U_i];
A2=[fai_11 I_03;I_04 fai_22];
N=a*A1*A2*A1';
N_11= A11*A2*A11';
[F2,G2]=eig(N);
G2;
N11=N(1:3,1:3);
N12=N(1:3,4:6);
N13=N(1:3,7:8);
N14=N(1:3,9:11);
N21=N(4:6,1:3);
N22=N(4:6,4:6);
N23=N(4:6,7:8);
N24=N(4:6,9:11);
N31=N(7:8,1:3);
N32=N(7:8,4:6);
N33=N(7:8,7:8);
N34=N(7:8,9:11);
N41=N(9:11,1:3);
N42=N(9:11,4:6);
N43=N(9:11,7:8);
N44=N(9:11,9:11);
P = sdpvar(3,3);
L = sdpvar(2,3);
P2 =P-(b*I_3);
L1 = [[  P2-N11        I_0-N12   I_05-N13   I_0-N14;
          I_0-N21       -P-N22    -(L')-N23   I_0-N24;
          I_06-N31      -L-N32    I_07-N33    L-N34; 
          I_0-N41       I_0-N42   (L')-N43    P-N44]>=0];
L2 = [P>=0]
LL =  L1+L2;
solvesdp(LL);
P1 = double(P);
[F,G]=eig(P1);
G
k = double(L)*inv(double(P))
U=A+(B*k);
[F1,E]=eig(U);
E
```

**Reply by johanlofberg** [reply]
There is no official command in YALMIP called solvesdp
https://yalmip.github.io/tutorial/basics/

If you are using the latest version in YALMIP, the warning is revealing the bug in your code

```
>> L1 = [[  P2-N11        I_0-N12   I_05-N13   I_0-N14;
          I_0-N21       -P-N22    -(L')-N23   I_0-N24;
          I_06-N31      -L-N32    I_07-N33    L-N34; 
          I_0-N41       I_0-N42   (L')-N43    P-N44]>=0];
Warning: Suspect non-symmetry in square constaint (Learn more)  
> In constraint (line 82)
  In  >=  (line 5) 
```

**Reply by yunjiandaiju** [reply]
```
 clear, clc;

A = [0.850 -0.038 -0.380; 
    0.735 0.815   1.594; 
    -0.664 0.697  -0.064];
B = [1.431 0.705;
    1.620 -1.129; 
    0.913 0.369];
u = randn(2,20);
x0 = randn(3, 1);

n = 3; % 
num_samples = 100; % 
desired_norm_square = 0.5; %
scaled_noises = zeros(n, num_samples);
for i = 1:num_samples
    noise = randn(n, 1); 
    while norm(noise, 2)^2 > desired_norm_square
        noise = randn(n, 1); 
    end
    scaled_noises(:, i) = noise;    
end
num_selected = 20;
selected_indices = randperm(num_samples, num_selected);
w= scaled_noises(:, selected_indices);
t=20;x=[];x(:,1)=x0;
for i=1:t
    x(:,i+1)=A*x(:,i)+B*u(:,i)+w(:,i);
end
x_i0=x(:,1:20);
x_i1=x(:,2:21);
U_i=u;
I_3=eye(3);
I_0=zeros(3,3);
I_01=zeros(2,3);
I_02=zeros(1,20);
I_03=zeros(3,20);
I_04=zeros(20,3);
I_05=zeros(3,2);
I_06=zeros(2,3);
I_07=zeros(2,2);
a=randn;
b=randn;
fai_11=t*0.5*eye(3);
fai_22=-eye(20); 
A1=[I_3,x_i1;I_0,-x_i0;I_01,-U_i;I_0,I_03];
A11=[I_3,x_i1;I_0,-x_i0;I_01,-U_i];
A2=[fai_11 I_03;I_04 fai_22];
N=a*A1*A2*A1';
N_11= A11*A2*A11';
[F2,G2]=eig(N);
G2;
N11=N(1:3,1:3);
N12=N(1:3,4:6);
N13=N(1:3,7:8);
N14=N(1:3,9:11);
N21=N(4:6,1:3);
N22=N(4:6,4:6);
N23=N(4:6,7:8);
N24=N(4:6,9:11);
N31=N(7:8,1:3);
N32=N(7:8,4:6);
N33=N(7:8,7:8);
N34=N(7:8,9:11);
N41=N(9:11,1:3);
N42=N(9:11,4:6);
N43=N(9:11,7:8);
N44=N(9:11,9:11);
yalmip('clear')
P = sdpvar(3,3);
L = sdpvar(2,3);
P2 =P-(b*I_3);
constraints=[[  P2-N11        I_0-N12   I_05-N13   I_0-N14;
          I_0-N21       -P-N22    -(L')-N23   I_0-N24;
          I_06-N31      -L-N32    I_07-N33    L-N34; 
          I_0-N41       I_0-N42   (L')-N43    P-N44]>=0]
Objective = [];
options = sdpsettings('verbose',1,'solver','quadprog','quadprog.maxiter',100);
% Solve the problem
sol = optimize(constraints,Objective,options);
P1 = double(P);
[F,G]=eig(P1);
G
k = double(L)*inv(double(P))
U=A+(B*k);
[F1,E]=eig(U);
E
```

**Reply by yunjiandaiju** [reply]
Is that the modification?

**Reply by johanlofberg** [reply]
No, it is clearly telling you that first constraint is unsymmetric which makes no sense as it is supposed to be an LMI. Also, trying to use quadprog to solve a semidefinite problem makes no sense either.

Finally, it can never ever be feasible even when you fix the violating non-symmetry, since I_07-N33 is negative definite

**Reply by yunjiandaiju** [reply]
```
clc;
clear all;
A = [0.850 -0.038 -0.380; 
0.735 0.815 1.594; 
-0.664 0.697 -0.064];
B = [1.431 0.705;
1.620 -1.129; 
0.913 0.369];
u = randn(2,20);
x0 = randn(3, 1);
%% 生成噪音
n = 3; % 噪声向量的长度
num_samples = 100; % 要生成的噪声向量数量
desired_norm_square = 0.5; %限制范数界
scaled_noises = zeros(n, num_samples);
for i = 1:num_samples
noise = randn(n, 1); % 生成高斯分布的随机噪声
while norm(noise, 2)^2 > desired_norm_square
noise = randn(n, 1); % 如果范数平方超过了阈值，重新生成噪声
end
scaled_noises(:, i) = noise; 
end
% 从生成的噪声中随机抽取20列
num_selected = 20;
selected_indices = randperm(num_samples, num_selected);
w= scaled_noises(:, selected_indices);
%% 生成响应
t=20;x=[];x(:,1)=x0;
for i=1:t
x(:,i+1)=A*x(:,i)+B*u(:,i)+w(:,i);
end
x_i0=x(:,1:20);
x_i1=x(:,2:21);
U_i=u;
%% 求解LMI
%定义给定矩阵
I_3=eye(3);
I_0=zeros(3,3);
I_01=zeros(2,3);
I_02=zeros(1,20);
I_03=zeros(3,20);
I_04=zeros(20,3);
I_05=zeros(3,2);
I_06=zeros(2,3);
I_07=zeros(2,2);
a=randn;
fai_11=t*0.5*eye(3);
fai_22=-eye(20); 
%% 定义N
A1=[I_3,x_i1;I_0,-x_i0;I_01,-U_i;I_0,I_03];
A11=[I_3,x_i1;I_0,-x_i0;I_01,-U_i];
A2=[fai_11 I_03;I_04 fai_22];
N=a*A1*A2*A1';
N11=N(1:3,1:3);
N12=N(1:3,4:6);
N13=N(1:3,7:8);
N14=N(1:3,9:11);
N21=N(4:6,1:3);
N22=N(4:6,4:6);
N23=N(4:6,7:8);
N24=N(4:6,9:11);
N31=N(7:8,1:3);
N32=N(7:8,4:6);
N33=N(7:8,7:8);
N34=N(7:8,9:11);
N41=N(9:11,1:3);
N42=N(9:11,4:6);
N43=N(9:11,7:8);
N44=N(9:11,9:11);

beita=normrnd(0,1);
I=eye(3);

% 定义 P 矩阵
P = sdpvar(3,3); 
L = sdpvar(2,3);  
 

%% 描述
Lmi_11=P-beita*eye(3);
Lmi_21=zeros(3,3);

Lmi_22=-P;
Lmi_31=zeros(2,3);
Lmi_32=-L;
Lmi_33=zeros(2,2);

Lmi_41=zeros(3,3);
Lmi_42=zeros(3,3);
Lmi_43=L';
Lmi_44=P;

cons =[Lmi_11,Lmi_21',Lmi_31',Lmi_41'';
         Lmi_21,Lmi_22,Lmi_32',Lmi_42';
         Lmi_31,Lmi_32,Lmi_33,Lmi_43';
         Lmi_41,Lmi_42,Lmi_43,Lmi_44]-[N11,N12,N13,N14;N21,N22,N23,N24;N31,N32,N33,N34;N41,N42,N43,N44]>=0;
      
 
%ops = sdpsettings('solver','MOSEK');
ops = sdpsettings('solver','SDPT3');
optimize(cons,[],ops);
```

**Reply by yunjiandaiju** [reply]
If that's the case what should I do with it, I'm trying to solve for this LMI and I'm not sure how I should go about it anymore or what method to use to solve for that LMI like in the picture I posted at the top, where only P and L are variables.

**Reply by johanlofberg** [reply]
That whole N-matrix is just numerical disaster. If you check it, you will see that it is not symmetric

However, even if you clean that up, the model is simply infeasible as LMI_33 is zero and the corresponding block in XX is psd, hence the that block in the diagonal is trivially negative definite so model can never be feasible

**Reply by yunjiandaiju** [reply]
So how should I solve for the LMI depicted in the picture? I have been confused about this for a long time. Because I was able to solve it in the above way, but the solution is incorrect.I don't know how I should go about solving it.

**Reply by yunjiandaiju** [reply]
Is it OK if I solve this LMI directly for the Schur complement about the fourth diagonal block? Because I found that the Schur complement of N with respect to the fourth diagonal block is a symmetric matrix.

**Reply by johanlofberg** [reply]
The problem is simply not solvable. You are asking for a positive number x which satisfies x<= -1. No matter how you write that, you are asking for the impossible

**Reply by yunjiandaiju** [reply]
If I ask for a solution to the problem that there exists a large matrix A with two variables (P,L), I require P to be real symmetric positive definite, and I also need this large matrix A to be semi-positive definite. How should I solve this kind of problem, and what type of problem does it belong to (semidefinite programming, linear programming, or something else?) , Can this kind of problem be solved?

**Reply by johanlofberg** [reply]
Very unclear what you are asking, "a large matrix A with two variables with (P,L)"??

**Reply by yunjiandaiju** [reply]
The problem I am trying to solve is that there exists a matrix A with two matrix variables (P,L), and it is required that P is real symmetric positive definite, and it is also required that this matrix A is semi-positive definite.

**Reply by johanlofberg** [reply]
So just like you have above, and that is how you code it. It is just that the problem you have stated isn't feasible

**Reply by yunjiandaiju** [reply]
I didn't understand exactly what you were referring to when you said it wasn't feasible. Can you help me explain in detail why this issue is not feasible? I ran him and it did prompt me the same way, but I don't know exactly what is causing it.
```
clc;
clear all;
A = [0.850 -0.038 -0.380; 
0.735 0.815 1.594; 
-0.664 0.697 -0.064];
B = [1.431 0.705;
1.620 -1.129; 
0.913 0.369];
u = randn(2,20);
x0 = randn(3, 1);
%% 生成噪音
n = 3; % 噪声向量的长度
num_samples = 100; % 要生成的噪声向量数量
desired_norm_square = 0.5; %限制范数界
scaled_noises = zeros(n, num_samples);
for i = 1:num_samples
noise = randn(n, 1); % 生成高斯分布的随机噪声
while norm(noise, 2)^2 > desired_norm_square
noise = randn(n, 1); % 如果范数平方超过了阈值，重新生成噪声
end
scaled_noises(:, i) = noise; 
end
% 从生成的噪声中随机抽取20列
num_selected = 20;
selected_indices = randperm(num_samples, num_selected);
w= scaled_noises(:, selected_indices);
%% 生成响应
t=20;x=[];x(:,1)=x0;
for i=1:t
x(:,i+1)=A*x(:,i)+B*u(:,i)+w(:,i);
end
x_i0=x(:,1:20);
x_i1=x(:,2:21);
U_i=u;
%% 求解LMI
%定义给定矩阵
I_3=eye(3);
I_0=zeros(3,3);
I_01=zeros(2,3);
I_02=zeros(1,20);
I_03=zeros(3,20);
I_04=zeros(20,3);
I_05=zeros(3,2);
I_06=zeros(2,3);
I_07=zeros(2,2);
a=randn;
fai_11=t*0.5*eye(3);
fai_22=-eye(20); 
%% 定义N
A1=[I_3,x_i1;I_0,-x_i0;I_01,-U_i];
A2=[fai_11 I_03;I_04 fai_22];
N=a*A1*A2*A1';
b=randn;
I=eye(3);
I_01=zeros(3,2);
I_02=zeros(3,3);
I_03=zeros(2,3);
I_04=zeros(3,5);
I_05=zeros(5,5);
b_I=b*I;
% 定义 P 矩阵
P = sdpvar(3,3); 
K = sdpvar(2,3);  
 %% 描述
Cons =[P,I_02,I_01;
         I_02,-P,-P*K';
         I_03,-K*P,-K*P*K']-N-[b_I,I_04;I_04',I_05]>=0;   
objective = [];
options = sdpsettings('solver','MOSEK');
sol = optimize(Cons,objective,options);
```

**Reply by johanlofberg** [reply]
You still haven't fixed your nonsymmetry 

>> [P,I_02,I_01;
         I_02,-P,-P*K';
         I_03,-K*P,-K*P*K']-N-[b_I,I_04;I_04',I_05]
Polynomial matrix variable 8x8 (full, real, 12 variables)
Coefficient range: 1 to 121519943.4535
>> 

Now even worse though, as you know have a polynomial model, hence no solver capable of solving the problem (after you would have fixed the nonsymmetry, and the horrible numerics you have)

**Reply by yunjiandaiju** [reply]
![image](https://github.com/yalmip/YALMIP/assets/142418156/942cef46-4ceb-40ab-ba62-523668405c75)
But I am programming according to this picture, where P,L are variables and the rest are known. All are symmetric as seen in the picture.And I'm matrix constrained for the whole, I don't understand why it's element constrained for the open.If my program is faulty, I would like to know how the LMI shown in the picture should be solved.
![image](https://github.com/yalmip/YALMIP/assets/142418156/b8c54086-043a-4c8f-818c-23b39355ba22)

**Reply by johanlofberg** [reply]
I already told you, it is not symmetric since your data isn't symmetric as you have horrible numerics

```
>> norm(N-N')

ans =

   9.5425e-09
```

You have K and P as decision variables, hence polynomial products in the model

**Reply by yunjiandaiju** [reply]
I feel like I'm a bit confused now, if only for the LMI shown in the picture, the variables are P,L. P is real symmetric positive definite. The rest of the letters are known, how do I write a program to solve for them. Can you help me a bit? Thank you very much.
![image](https://github.com/yalmip/YALMIP/assets/142418156/f62c762e-6b21-48a8-855c-6c04372694eb)

**Reply by johanlofberg** [reply]
Look at the code you just posted, it is not what you are showing a picture of

**Reply by yunjiandaiju** [reply]
Yes, the code just now is the Schur complement of the fourth diagonal block of this image. I thought a lot about how to fix it, so I wrote the code above. Here is my code for the image.
```
clc;
clear all;
A = [0.850 -0.038 -0.380; 
0.735 0.815 1.594; 
-0.664 0.697 -0.064];
B = [1.431 0.705;
1.620 -1.129; 
0.913 0.369];
u = randn(2,20);
x0 = randn(3, 1);
%% 生成噪音
n = 3; % 噪声向量的长度
num_samples = 100; % 要生成的噪声向量数量
desired_norm_square = 0.5; %限制范数界
scaled_noises = zeros(n, num_samples);
for i = 1:num_samples
noise = randn(n, 1); % 生成高斯分布的随机噪声
while norm(noise, 2)^2 > desired_norm_square
noise = randn(n, 1); % 如果范数平方超过了阈值，重新生成噪声
end
scaled_noises(:, i) = noise; 
end
% 从生成的噪声中随机抽取20列
num_selected = 20;
selected_indices = randperm(num_samples, num_selected);
w= scaled_noises(:, selected_indices);
%% 生成响应
t=20;x=[];x(:,1)=x0;
for i=1:t
x(:,i+1)=A*x(:,i)+B*u(:,i)+w(:,i);
end
x_i0=x(:,1:20);
x_i1=x(:,2:21);
U_i=u;
%% 求解LMI
%定义给定矩阵
I_3=eye(3);
I_0=zeros(3,3);
I_01=zeros(2,3);
I_02=zeros(1,20);
I_03=zeros(3,20);
I_04=zeros(20,3);
I_05=zeros(3,2);
I_06=zeros(2,3);
I_07=zeros(2,2);
a=abs(randn);
fai_11=t*0.5*eye(3);
fai_22=-eye(20); 
%% 定义N
A1=[I_3,x_i1;I_0,-x_i0;I_01,-U_i;I_0,I_03];
A11=[I_3,x_i1;I_0,-x_i0;I_01,-U_i];
A2=[fai_11 I_03;I_04 fai_22];
N_1=A11*A2*A11';
N=a*A1*A2*A1';
N11=N(1:3,1:3);
N12=N(1:3,4:6);
N13=N(1:3,7:8);
N14=N(1:3,9:11);
N21=N(4:6,1:3);
N22=N(4:6,4:6);
N23=N(4:6,7:8);
N24=N(4:6,9:11);
N31=N(7:8,1:3);
N32=N(7:8,4:6);
N33=N(7:8,7:8);
N34=N(7:8,9:11);
N41=N(9:11,1:3);
N42=N(9:11,4:6);
N43=N(9:11,7:8);
N44=N(9:11,9:11);
beita=abs(randn);
I=eye(3);
% 定义 P 矩阵
P = sdpvar(3,3); 
L = sdpvar(2,3); 
Z = sdpvar(5,3);
 %% 描述
Lmi_11=P-beita*eye(3);
Lmi_21=zeros(3,3);
Lmi_22=-P;
Lmi_31=zeros(2,3);
Lmi_32=-L;
Lmi_33=zeros(2,2);
Lmi_41=zeros(3,3);
Lmi_42=zeros(3,3);
Lmi_43=L';
Lmi_44=P;
Cons =[Lmi_11,Lmi_21',Lmi_31',Lmi_41;
         Lmi_21,Lmi_22,Lmi_32',Lmi_42;
         Lmi_31,Lmi_32,Lmi_33,Lmi_43';
         Lmi_41,Lmi_42,Lmi_43,Lmi_44]-[N11,N12,N13,N14;N21,N22,N23,N24;N31,N32,N33,N34;N41,N42,N43,N44]>=0;   
objective = [norm(P)];
options = sdpsettings('solver','MOSEK');
sol = optimize(Cons,objective,options);
% ops = sdpsettings('solver','MOSEK');
% ops = sdpsettings('solver','SDPT3');
% optimize(cons,[],ops);
 P=value(P);
 L=value(L);
 k=L*inv(P);
KK=A+B*k;
[F,G]=eig(KK);
 G

```

**Reply by yunjiandaiju** [reply]
My goal was to fix the LMI in the picture,so I thought of a number of ways to do it,including his Schur patch. Because I was confused for too long leading to maybe a bit of confusion. My ultimate goal is to solve the LMI in the picture.
![image](https://github.com/yalmip/YALMIP/assets/142418156/1e6586af-a770-45af-af61-cbaaa571169d)

**Reply by johanlofberg** [reply]
I repeat for the third time: **Your data is not symmetric**

```
>> N = [N11,N12,N13,N14;N21,N22,N23,N24;N31,N32,N33,N34;N41,N42,N43,N44];
>> norm(N-N')

ans =

   5.8604e-09

```

**Reply by yunjiandaiju** [reply]
I see what you mean, but mathematically shouldn't N be symmetric? Is this due to the fact that I'm giving too large a value? The difference in Frobenius norms between the matrix N and its transpose N' was calculated and the result is 5.8604e-09. this is very close to zero, but not exactly zero.

**Reply by johanlofberg** [reply]
As I said, you have horrible numerics, and thus have to fix that to ensure you actually have symmetric data

**Reply by yunjiandaiju** [reply]
So it's caused by the presence of very large numbers in N. If I fix this, he should be symmetric.

**Reply by johanlofberg** [reply]
Large, small, trickle.flow or what ever which is causing numerical issues.

I hope this isn't a surprise to you that floating-point numbers have errors

```
>> x = sqrt(2);x^2-2

ans =

   4.4409e-16
```

Since you know the matrix should be symmetric and it is just numerical noise, just symmetrize it afterwards (although best is of course if you can understand why you have crap data, since crap data in an SDP model will most likely cause numerical problems once the solver tries to solve it. crap in crap out)

**Reply by yunjiandaiju** [reply]
Okay, thank you Professor. I have figured out what my problem was, thank you very much.
